runtime/sema/type.go

```diff
@@ -3180,6 +3221,42 @@ func (t *FunctionType) Resolve(typeArguments *TypeParameterTypeOrderedMap) Type
 
 }
 
+func (t *FunctionType) Map(gauge common.MemoryGauge, f func(Type) Type) Type {
+	returnType := t.ReturnTypeAnnotation.Map(gauge, f)
+
+	var newParameters []Parameter = make([]Parameter, 0, len(t.Parameters))
+	for _, parameter := range t.Parameters {
+		newParameterTypeAnnot := parameter.TypeAnnotation.Map(gauge, f)
+
+		newParameters = append(
+			newParameters,
+			Parameter{
+				Label:          parameter.Label,
+				Identifier:     parameter.Identifier,
+				TypeAnnotation: newParameterTypeAnnot,
+			},
+		)
+	}
+
+	var newTypeParameters []*TypeParameter = make([]*TypeParameter, 0, len(t.TypeParameters))
+	for _, parameter := range t.TypeParameters {
+		newTypeParameterTypeBound := parameter.TypeBound.Map(gauge, f)
+
+		newTypeParameters = append(
+			newTypeParameters,
+			&TypeParameter{
+				Name:      parameter.Name,
+				Optional:  parameter.Optional,
+				TypeBound: newTypeParameterTypeBound,
+			},
```

Comment:
These new type parameters, or actually, the mapping of old to new type parameter, should probably be passed along in the recursive map calls, so that generic type occurrences can be "fixed up"